File: ios/Reguerta/Reguerta/Data/Calendar/ChainedDeliveryCalendarRepository.swift
import Foundation

struct ChainedDeliveryCalendarRepository<Primary: DeliveryCalendarRepository, Fallback: DeliveryCalendarRepository>:
    DeliveryCalendarRepository {
    let primary: Primary
    let fallback: Fallback

    func defaultDeliveryDayOfWeek() async throws -> DeliveryWeekday {
        try await primary.defaultDeliveryDayOfWeek()
    }

    func allOverrides() async throws -> [DeliveryCalendarOverride] {
        try await primary.allOverrides()
    }

    func upsertOverride(_ override: DeliveryCalendarOverride) async throws -> DeliveryCalendarOverride {
        try await primary.upsertOverride(override)
    }

    func deleteOverride(weekKey: String) async throws {
        try await primary.deleteOverride(weekKey: weekKey)
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/data/calendar/ChainedDeliveryCalendarRepository.kt
package com.reguerta.user.data.calendar

import com.reguerta.user.domain.calendar.DeliveryCalendarOverride
import com.reguerta.user.domain.calendar.DeliveryCalendarRepository
import com.reguerta.user.domain.calendar.DeliveryWeekday

class ChainedDeliveryCalendarRepository(
    private val primary: DeliveryCalendarRepository,
    @Suppress("UNUSED_PARAMETER") fallback: DeliveryCalendarRepository,
) : DeliveryCalendarRepository {
    override suspend fun getDefaultDeliveryDayOfWeek(): DeliveryWeekday? =
        primary.getDefaultDeliveryDayOfWeek()

    override suspend fun getAllOverrides(): List<DeliveryCalendarOverride> = primary.getAllOverrides()

    override suspend fun upsertOverride(override: DeliveryCalendarOverride): DeliveryCalendarOverride =
        primary.upsertOverride(override)

    override suspend fun deleteOverride(weekKey: String) = primary.deleteOverride(weekKey)
}
